### ISSUE TITLE
Pinky and Ring fingers bend when hand rotates, how can I disable this?

### ISSUE THREAD

**Issue by guiglass** [ValveSoftware/steamvr_unity_plugin#554]
Hi, I notice whenever I rotate my knuckles controllers around my wrist I see the pinky/ring finger of the VR hand are bending a little as if reacting to gravity, i think? but this has very undesirable effects to my player's fingers in my game and it sometimes appears that I'm not holding things correctly.

can someone PLEASE help me turn off the pinky/ring finger gravity bending feature? Please!!!
Thank you 👍 👍

**Comment by guiglass** [comment]
Has anyone ever solved this?

Did anyone at Steam ever think that feature might cause huge headaches for developer trying to create only simple VR hands?

Why is it so hard to TURN IT OFF??????

**Comment by guiglass** [comment]
I see a new bug, appears that Steam didn't even invert the bend angles properly. 
When i face left hand palm down, left fingers bend inward. 
When i face right hand palm up, right fingers bend inward.  

Why does facing my right palm up make fingers bind indwards?? This finger bending feature is obviously super bugged and extremely annoying.

I have presented video evidence proving that even the Steam example is broken.
**Can someone please tag this bending issue as a confirmed bug so someone can start fixing it?**

**Comment by guiglass** [comment]
Anyone?

**Comment by airmouse** [comment]
Yep!! Having same exact issue here. There appears to be no fix or any way to turn off the finger curling features.

And good catch on noticing the Left and Right hand are not mirrored correctly. Now that I am looking for it I can say it makes the VR hand look super fake!!

Whoever wrote the  skeleton poser stuff definitely did not test if very well.

If Valve would allow us to disable the gravity influence (or as you call it: user wrist rotation) then all problems would be solved.

Instead they added some useless extra feature and essentially made the whole thing crap.

Please fix this valve, I don't want my fingers curling by themselves. WTF?

**Comment by airmouse** [comment]
Yep!! Having same exact issue here. There appears to be no fix or any way to turn off the finger curling feature.

And good catch on noticing the Left and Right hand are not mirrored correctly. Now that I am looking for it I can clearly say it makes the VR hand look super fake!!

Whoever wrote the skeleton poser stuff definitely did not test if very well.

If Valve only would allow us to disable the gravity influence (or as you call it: user wrist rotation) then all problems would be solved.

Instead they just added some useless extra feature and essentially made the whole thing crap.

Please fix this Valve, I don't want my fingers curling by themselves. WTF?

**Comment by airmouse** [comment]
Yep!! Having same exact issue here. There appears to be no fix and now way to disable the finger curling feature.

And good catch on noticing the Left and Right hand are not mirrored correctly. Now that I am looking for it I can clearly say it makes the VR hand look super fake!!

Whoever wrote the skeleton poser stuff definitely did not test if very well.

If Valve only would allow us to disable the gravity influence (or as you call it: user wrist rotation) then all problems would be solved.

Instead they just added some useless extra feature and essentially made the whole thing crap.

Please fix this Valve, I don't want my fingers curling by themselves. WTF?

**Comment by airmouse** [comment]
Yep!! Having same exact issue here. There appears to be no fix and no way to turn off the finger curling feature.

And good catch on noticing the Left and Right hand are not mirrored correctly. Now that I am looking for it I can clearly say it makes the VR hand look super fake!!

Whoever wrote the skeleton poser stuff definitely did not test if very well.

If Valve only would allow us to disable the gravity influence (or as you call it: user wrist rotation) then all problems would be solved.

Instead they just added some useless extra feature and essentially made the whole thing crap.

Please fix this Valve, I don't want my fingers curling by themselves. WTF?

**Comment by zite** [comment]
Hi, can you talk a little more about your specific scenarios so we can better understand the situation? What are you trying to do with the skeletal data that is negatively effected by this behavior?

I can talk a little bit about why this decision was made. The purpose of Skeletal Input is to get our best estimation of what the users hands are doing. With vive wands we don't get a lot of data but we notice people's fingers tend to curl slightly when they raise them inwards and relax out when they put them down so we were replicating this behavior in the Without_Controller data.

It may be helpful to switch to the With_Controller data to get more accurate finger positions but they'll look like they're curled around a controller.

**Comment by guiglass** [comment]
My specific scenario is exactly the same as the Alien floppy hands example.
I followed **exactly** the Alien floppy hands example provided in this Git. I created a set of custom hands and linked them to a proxy skeleton, this is the exact same way the alien floppy hands example was done in your example (which I download from this Git).

Because I followed your example **exactly** I get **exactly** the same bug that exists in your Alien floppy hand's example:
https://youtu.be/VYil9uY7kds?t=4

Do you not see the alien floppy finger curls up weirdly?? Did no one ever notice this? I think it's clear as day,

If you need more information then I implore you to please what the video, then try alien floppy hands example again, I am confident you will notice the strange curling issue if you look for it.

It is all caused by the pinky curly nonsens that causes all these freakin issues. I understand you were attempting to mimic real-life. But by doing so you made it nearly impossible to create custom hands.

I think it was very unfair of you to promoted the alien floppy hands example leaving without properly addressing that curing issue.

My game design is already locked in, and you will probably never add options for turning of the finger curing gravity features. Which means no matter how much I try, my custom hands will alwasy suffer from weird curling issues (just like the alien floppy hands).

I just want to turn off the auto finger curing, PLEASE!!!!

**Comment by zite** [comment]
The alien hands in the example shouldn't be curling quite that much. If they only curled slightly (like they do in the normal example hands) would that work for you?

**Comment by guiglass** [comment]
Hi, I am not sure if I follow you, are you saying that you can see the curling issue, but that it is not as severe for you than what I show in the video?

I don't understand how that can be, because I recently downloaded this Git repo and right out of the box the alien floppy hands are curling nearly 90 degrees when I flip my wrist over.

I also tested the demo on two machines (Win7 and Win10) and both were exactly the same.

Maybe you have different settings than I do, if so, what are they? Because I would really like to get rid of the auto-curling feature as much as possible.

**Comment by zite** [comment]
Sorry I wasn't clear, I meant that it is a bug that they are curling that much. I don't believe the normal hands curl that much. The alien hands are a unity-specific sample that may have a bug that the normal hands do not have.

If you're looking to do your own curl based animations separate from our estimation you may want to create your own vector1 actions for curls and bind those per controller. The point of Skeletal Input is to give you our best guess at finger positions.

We have a specific api you can use for curls specifically: SteamVR_Action_Skeleton has a summaryDataType member (https://github.com/ValveSoftware/steamvr_unity_plugin/blob/a471e5a94c2ea823b79654616cff0c54453b1eff/Assets/SteamVR/Input/SteamVR_Action_Skeleton.cs#L114) that you can set to "FromDevice" which will give you data direct from the device. But only if the device provides it. If it doesn't then we'll substitute the FromAnimation data. 

For scenarios where you really do care about curls you probably want to customize what triggers a curl on a specific controller. Which you can do with actions in the binding layer. If you want to leave to up to the controller _only_ when it _has_ per finger data you can check SteamVR_Action_Skeleton's skeletalTrackingLevel (https://github.com/ValveSoftware/steamvr_unity_plugin/blob/a471e5a94c2ea823b79654616cff0c54453b1eff/Assets/SteamVR/Input/SteamVR_Action_Skeleton.cs#L126)

**Comment by guiglass** [comment]
That's cool, and sounds very promising but I am unsure how to use it.

I just tried adding a script that runs after the scene has fully loaded. 

I try to set the summaryDataType for all hands in the scene, but I did not see any of them change. Am I using it correctly?
```

    void OnEnable()
    {
        foreach (var hand in GameObject.FindObjectsOfType<SteamVR_Behaviour_Skeleton>())
        {
            //set the summaryDataType for each SteamVR_Behaviour_Skeleton in the scene.
            hand.skeletonAction.summaryDataType = EVRSummaryType.FromAnimation;
        }
    }
```
What should I expect to see after changing that value? I would expect the skeleton to immediately start using raw values from the device, but I am very unsure.

And I truly do appreciate your assistance.

**Comment by guiglass** [comment]
I still an not sure how to implement the solution into my game.

When do I call fingerCurls ? What should I supply to disable curling of gravity influence?

What is the expected result of change fingerCurls array?

Are there any examples showing it's usage? Or any documentation?

My characters fingers are still bending strangely and it is a super let down to know there might be a solution, but that it's too vague to understand how to use it.

**Comment by guiglass** [comment]
Can someone please help me understand how to call your example code? I am still at a loss. Need help please!

**Comment by guiglass** [comment]
Obviously this bug is never going to be fixed. And no one is giving it any attention. 

I'm not trying to be rude but your SDK is very broken Steam.

**Comment by guiglass** [comment]
I tried using your example, but the fingers still curl. Obviously not a solution, Still seeking an answer to how to prevent fingers from curling when wrist is rotated.
